doc/coq.rst

```diff
@@ -0,0 +1,392 @@
+
+- ``(prelude <name>)`` refers to the Coq source that contains the extraction
+  commands.
+
+- ``(extracted_modules <names>)`` is an exhaustive list of OCaml modules
+  extracted.
+
+- ``<optional-fields>`` are ``flags``, ``theories``, and ``libraries``. All of
+  these fields have the same meaning as in the ``coq.theory`` stanza.
+
+The extracted sources can then be used in ``executable`` or ``library`` stanzas
+as any other sources.
+
+Note that the sources are extracted to the directory where the ``prelude`` file
+lives. Thus the common placement for the ``OCaml`` stanzas is in the same
+:ref:`dune file<dune-files>`.
+
+**Warning**: using Coq's ``Cd`` command to work around problems with the output
+directory is not allowed when using extraction from Dune. Moreover the ``Cd``
+command has been deprecated in Coq 8.12.
+
+.. _coq-pp:
+
+coq.pp
+------
+
+Authors of Coq plugins often need to write ``.mlg`` files to extend the Coq
+grammar. Such files are preprocessed with the ``coqpp`` binary. To help plugin
+authors avoid writing boilerplate, we provide a ``(coqpp ...)`` stanza:
+
+.. code:: scheme
+
+    (coq.pp (modules <mlg_list>))
+
+which, for each ``g_mod`` in ``<mlg_list>``, is equivalent to the following
+rule:
+
+.. code:: lisp
+
+    (rule
+     (targets g_mod.ml)
+     (deps (:mlg-file g_mod.mlg))
+     (action (run coqpp %{mlg-file})))
+
+.. _examples:
+
+Examples of Coq Projects
+------------------------
+
+Here we list some examples of some basic Coq project setups in order.
+
+.. _example-simple:
+
+Simple Coq Project
+~~~~~~~~~~~~~~~~~~
+
+Let us start with a simple project. First we make sure we have a
+:ref:`dune-project` file with a :ref:`Coq lang<coq-lang>` stanza present:
+
+.. code:: scheme
+
+  (lang dune 3.2)
+  (using coq 0.3)
+
+Next we need a :ref:`dune file<dune-files>` with a :ref:`coq-theory` stanza:
+
+.. code:: scheme
+  
+  (coq.theory
+   (name myTheory))
+
+
+Finally, we need a Coq ``.v`` file which we name ``A.v``:
+
+
+.. code:: coq
+
+  (** This is my def *)
+  Definition mydef := nat.
+
+Now we run ``dune build``. After this is complete, we get the following files:
+
+.. code::
+
+  .
+  ├── A.v
+  ├── _build
+  │   ├── default
+  │   │   ├── A.glob
+  │   │   ├── A.v
+  │   │   ├── A.v.d
+  │   │   └── A.vo
+  │   └── log
+  ├── dune
+  └── dune-project
+
+.. _example-multi-theory:
+
+Multi-Theory Project
+~~~~~~~~~~~~~~~~~~~~
+
+Here is an example of a more complicated setup:
+
+.. code::
+
+  .
+  ├── A
+  │   ├── AA
+  │   │   └── aa.v
+  │   ├── AB
+  │   │   └── ab.v
+  │   └── dune
+  ├── B
+  │   ├── b.v
+  │   └── dune
+  └── dune-project
+
+Here are the :ref:`dune files<dune-files>`:
+
+.. code:: scheme
+
+  ; A/dune
+  (include_subdirs qualified)
+  (coq.theory
+   (name A))
+
+  ; B/dune
+  (coq.theory
+   (name B)
+   (theories A))
+
+Notice the ``theories`` field in ``B`` allows one :ref:`coq-theory` to depend on
+another. Another thing to note is the inclusion of the :ref:`include_subdirs`
```

Comment:
How do users identify the dependencies and update their dune files?  Will there be tooling to do that automatically?  I'm not sure how significant an issue this may be for users.

Reply:
That's a bit of an art, and likely Dune won't be able to help a lot until 2.0 coq lang.

Reply:
I thought it would be an update to coqdep.  Indeed, it might be cool if fine could read and apply coqdep output.  IIUC.

Reply:
Oh, I misunderstood you. Dune of course calls coqdep to find dependencies of .v files, `(theories ...)` is for dependency on theories, that is to say, to put in scope whole set of compiled files into a prefix, that is to say, `-Q`.

Now, dune can't help a lot the user to know where a particular module is in terms of theories/packages.

Reply:
The idea is to enhance coqdoc so it can tell you what top-level modules/packages it depends on--the bits that need to appear in -Q options.  It could look at _CoqProject files and/or in the Coq loadpath.  (handwaving here)